core/src/main/java/io/undertow/server/DefaultByteBufferPool.java

```diff
@@ -256,20 +256,26 @@ private static class DefaultPooledBuffer implements PooledByteBuffer {
 
         @Override
         public ByteBuffer getBuffer() {
+            final ByteBuffer tmp = this.buffer;
             if(referenceCount == 0) {
                 throw UndertowMessages.MESSAGES.bufferAlreadyFreed();
             }
-            return buffer;
+            return tmp;
         }
 
         @Override
         public void close() {
-            if(referenceCountUpdater.compareAndSet(this, 1, 0)) {
-                if(leakDetector != null) {
+            final ByteBuffer tmp = this.buffer;
+            if (referenceCountUpdater.compareAndSet(this, 1, 0)) {
+                this.buffer = null;
+                if (leakDetector != null) {
                     leakDetector.closed = true;
                 }
-                pool.freeInternal(buffer);
-                this.buffer = null;
+                if (tmp != null) {
+                    pool.freeInternal(tmp);
+                    tmp.position(0);
+                    tmp.limit(0);
```

Comment:
I don’t think it’s safe to mutate position and limit after calling freeInternal, the buffer may already be borrowed elsewhere

Reply:
I'll elaborate a bit more on this. The risk of it being borrowed elsewhere is the same, regardless of whether we clear the buffer or not. If it is borrowed elsewhere and the close is invoked concurrently with the getBuffer mehod, we risk the following order:
* thread 1 gets the buffer
* thread 2 closes the pooled buffer and buffer is made available to a third thread
* thread 3 gets the buffer
* thread 1 uses the buffer
* thread 3 is doomed to see an error when it uses the buffer
So the risk exists regardless.
On the other hand, a normal reusage of the buffer implies that the buffer should be cleared before being reused anyway, so we must clear it here

Reply:
Ok, I just noticed that the clear is being done before returning the pooled byte buffer at allocate. Well, in this case, we can clear it at either one or the other point, either at close or at allocate, the effect will be the same, and in both cases we will have an inconsistency in thread1 or thread3 in the scenario I described above